docs/features/string-literals-data-section.md

```diff
@@ -0,0 +1,162 @@
+When the limit is reached, the following compiler error is reported by Roslyn:
+
+```
+error CS8103: Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string literals.
+```
+
+By turning on [the feature flag](#configuration), string literals ([where possible](#eligible-string-literals)) are instead emitted as UTF-8 data into a different section of the PE file
+which does not have the same limit. The emit format is similar to [explicit u8 string literals][u8-literals].
+
+The feature is currently implemented only for C#, not VB.
+
+## Configuration
+
+The feature flag can take a non-negative integer threshold. Only string literals whose length is over the threshold are emitted using the utf8 encoding strategy.
+By default, the threshold is 100. Specifying 0 means all string literals are considered for the feature. Specifying `off` turns off the feature (this is the default).
+
+The feature flag can be specified on the command line like `/features:data-section-string-literals` or `/features:data-section-string-literals=20`,
+or in a project file in a `<PropertyGroup>` like `<Features>$(Features);data-section-string-literals</Features>` or `<Features>$(Features);data-section-string-literals=20</Features>`.
+
+## Eligible string literals
+
+A string literal is emitted with the utf8 encoding if and only if all of these are true:
+- it can be encoded as UTF-8, i.e., the following call does not throw:
+  ```cs
+  new System.Text.Encoding.UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true).GetBytes(literal)
+  ```
+- its `Length` is greater than the threshold specified by [the feature flag](#configuration).
+
+## Emit
+
+The standard string literal emit strategy emits `ldstr "string"`.
+That puts the string literal on the UserString heap of the PE file.
+
+The utf8 string literal encoding emit strategy emits `ldsfld` of a field in a generated class instead.
+
+For every string literal, a unique internal static class is generated which
+- has name composed of `<S>` followed by a hex-encoded SHA-256 hash of the string,
+- lives in the global namespace,
+- has one `.data` field which is emitted the same way as for [explicit u8 string literals][u8-literals],
+- has one `string` field which is initialized in a static constructor of the class.
+
+The initialization calls `<PrivateImplementationDetails>.BytesToString` helper which in turn calls `Encoding.UTF8.GetBytes`.
+This is an optimization so each of the generated class static constructors is slightly smaller in IL size.
+These size savings can add up since one class is generated per one eligible string literal.
+
+The following example demonstrates the code generated for string literal `"Hello."`.
+
+```cs
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+[CompilerGenerated]
+internal static class <S>2D8BD7D9BB5F85BA643F0110D50CB506A1FE439E769A22503193EA6046BB87F7
```

Comment:
Would it be possible to avoid having one type per string?
We can have all the members for a given string (the data field, the `string` field (lazily initialized) and possibly a helper method to handle the lazy initialization) in `<PrivateImplementationDetails>`? Those three members can include the hash in their name to avoid conflicts.

Reply:
Also, by caching strings in a static field, are we creating a problem for the GC? Would using `WeakReference` solve that problem?

Reply:
> Would it be possible to avoid having one type per string?
> We can have all the members for a given string (the data field, the `string` field (lazily initialized) and possibly a helper method to handle the lazy initialization) in `<PrivateImplementationDetails>`? Those three members can include the hash in their name to avoid conflicts.

That's a possible alternative, I can mention it.

As @davidwrighton said elsewhere, the static constructor approach emits a single mov instruction and no branches and calls.

> Also, by caching strings in a static field, are we creating a problem for the GC?

It's not any different than current `string` literal behavior I think.

Reply:
Thanks. I just caught up on a Teams discussion you forwarded me today. It looks like the trade-off was discussed and the runtime team is not too worried about the proliferation of types, they could do some optimization and we're engaged on benchmarking.

Reply:
> [by caching strings in a static field, are we creating a problem for the GC?] It's not any different than current string literal behavior I think.

Let's confirm with runtime folks. I think we're changing the current string literal behavior.
Today, if I use `M("hello")` we do a `ldstr` with a metadata token to the string. This allocates for the string (unless the string is re-used/interned then we can reference existing allocated string) and later on that can be garbage collected.
With a static field, I don't think it can be collected unless the type is unloaded.

Reply:
I see you already have a follow-up tracked below in "GC" section. Thanks